File: src/ui/growth-display.ts
import type { Session, SkillScores, GrowthProfile, Quote, Trend } from '../types.js';

const FILLED = '█';
const EMPTY = '░';
const BAR_LENGTH = 10;

const RESET = '\x1b[0m';
const BOLD = '\x1b[1m';
const DIM = '\x1b[2m';
const GREEN = '\x1b[32m';
const RED = '\x1b[31m';
const YELLOW = '\x1b[33m';

function renderBar(score: number): string {
  const clamped = Math.max(1, Math.min(10, Math.round(score)));
  return FILLED.repeat(clamped) + EMPTY.repeat(BAR_LENGTH - clamped);
}

function scoreLabel(score: number): string {
  if (score >= 8) return 'strong';
  if (score >= 5) return 'moderate';
  return 'weak';
}

function trendIndicator(trend: Trend, delta?: number): string {
  if (trend === 'improving') {
    const d = delta !== undefined ? ` (+${delta})` : '';
    return `${GREEN}improving${d}${RESET}`;
  }
  if (trend === 'declining') {
    const d = delta !== undefined ? ` (${delta})` : '';
    return `${RED}declining${d}${RESET}`;
  }
  return `${DIM}stable${RESET}`;
}

export function renderGrowthScore(
  session: Session,
  growthProfile?: GrowthProfile,
  closingQuote?: Quote,
): string {
  const lines: string[] = [];

  const outcomeLabel = session.outcome === 'solved' ? 'Solved' : 'Abandoned';
  lines.push(
    `\n${BOLD}Session Complete — ${session.questionsToRootCause} questions to root cause (${outcomeLabel})${RESET}`,
  );
  lines.push('');
  lines.push(`${BOLD}Your Debugging Profile:${RESET}`);

  const scores = session.skillScores;
  const dims: Array<{ name: string; key: keyof SkillScores; label: string }> = [
    { name: 'Assumption-checking', key: 'assumptionChecking', label: scoreLabel(scores.assumptionChecking) },
    { name: 'Evidence-gathering', key: 'evidenceGathering', label: scoreLabel(scores.evidenceGathering) },
    { name: 'Root cause speed', key: 'rootCauseSpeed', label: scoreLabel(scores.rootCauseSpeed) },
  ];

  for (let i = 0; i < dims.length; i++) {
    const dim = dims[i];
    const bar = renderBar(scores[dim.key]);
    const connector = i < dims.length - 1 ? '┣' : '┗';
    let line = `${connector} ${dim.name.padEnd(22)} ${bar}  ${dim.label}`;

    if (growthProfile) {
      const trend = growthProfile.recentTrend[dim.key];
      if (trend !== 'stable') {
        line += ` — ${trendIndicator(trend)}`;
      }
    }

    lines.push(line);
  }

  if (growthProfile && growthProfile.totalSessions > 1) {
    lines.push('');
    lines.push(
      `${DIM}${growthProfile.totalSessions} sessions total | ${growthProfile.solvedCount} solved | ${growthProfile.abandonedCount} abandoned${RESET}`,
    );
  }

  if (closingQuote) {
    lines.push('');
    lines.push(`${DIM}"${closingQuote.text}"${RESET}`);
    lines.push(`${DIM} — ${closingQuote.philosopher}${RESET}`);
  }

  lines.push('');
  return lines.join('\n');
}

File: nietzschian_debugger/ui/growth_display.py
"""Growth score display for session completion."""

from __future__ import annotations

from typing import Optional

from ..types import GrowthProfile, Quote, Session, SkillScores, Trend

FILLED = "\u2588"
EMPTY = "\u2591"
BAR_LENGTH = 10

RESET = "\x1b[0m"
BOLD = "\x1b[1m"
DIM = "\x1b[2m"
GREEN = "\x1b[32m"
RED = "\x1b[31m"
YELLOW = "\x1b[33m"


def _render_bar(score: int) -> str:
    """Render a Unicode bar chart for a score (1-10)."""
    clamped = max(1, min(10, round(score)))
    return FILLED * clamped + EMPTY * (BAR_LENGTH - clamped)


def _score_label(score: int) -> str:
    """Return a text label for a score."""
    if score >= 8:
        return "strong"
    if score >= 5:
        return "moderate"
    return "weak"


def _trend_indicator(trend: Trend, delta: Optional[int] = None) -> str:
    """Render a colored trend indicator."""
    if trend == "improving":
        d = f" (+{delta})" if delta is not None else ""
        return f"{GREEN}improving{d}{RESET}"
    if trend == "declining":
        d = f" ({delta})" if delta is not None else ""
        return f"{RED}declining{d}{RESET}"
    return f"{DIM}stable{RESET}"


def render_growth_score(
    session: Session,
    growth_profile: Optional[GrowthProfile] = None,
    closing_quote: Optional[Quote] = None,
) -> str:
    """Render the growth score display for a completed session."""
    lines: list[str] = []

    outcome_label = "Solved" if session.outcome == "solved" else "Abandoned"
    lines.append(
        f"\n{BOLD}Session Complete -- {session.questions_to_root_cause} questions "
        f"to root cause ({outcome_label}){RESET}"
    )
    lines.append("")
    lines.append(f"{BOLD}Your Debugging Profile:{RESET}")

    scores = session.skill_scores
    dims = [
        ("Assumption-checking", scores.assumption_checking, "assumption_checking"),
        ("Evidence-gathering", scores.evidence_gathering, "evidence_gathering"),
        ("Root cause speed", scores.root_cause_speed, "root_cause_speed"),
    ]

    for i, (name, score, key) in enumerate(dims):
        bar = _render_bar(score)
        connector = "\u2523" if i < len(dims) - 1 else "\u2517"
        line = f"{connector} {name:<22} {bar}  {_score_label(score)}"

        if growth_profile:
            trend = getattr(growth_profile.recent_trend, key)
            if trend != "stable":
                line += f" -- {_trend_indicator(trend)}"

        lines.append(line)

    if growth_profile and growth_profile.total_sessions > 1:
        lines.append("")
        lines.append(
            f"{DIM}{growth_profile.total_sessions} sessions total | "
            f"{growth_profile.solved_count} solved | "
            f"{growth_profile.abandoned_count} abandoned{RESET}"
        )

    if closing_quote:
        lines.append("")
        lines.append(f'{DIM}"{closing_quote.text}"{RESET}')
        lines.append(f"{DIM} -- {closing_quote.philosopher}{RESET}")

    lines.append("")
    return "\n".join(lines)
